### PR TITLE
REF/BLD: migrate to latest standards (pyproject.toml)

### PR DESCRIPTION
**Pull request by tangkong**
## Description
Been a while since we've seen one of these huh?

- Migrates packaging framework to pyproject.toml, removing versioneer
- adds and runs pre-commit checks  (flake8, pycln, isort, etc)

## Context
Deprecated versioneer calls were preventing the installation of this package on py3.12

## Testing
```
$ conda create -n py312 python=3.12
$ conda activate py312
$ <clone pyca>
$ pip install .
$ pip install pytest, pcaspy  # test requirements

$ pytest
$ python -c "import psp, pyca; print(psp.Pv); print(pyca.capv)"

```

### CODE REVIEW

File: conda-recipe/meta.yaml [hunk 1]
```diff
@@ -1,33 +1,41 @@
...
 requirements:
   build:
-    - python {{ PY_VER }}*
-    - setuptools
-    - epics-base {{ EPICS }}*
-    - numpy {{ NPY_VER }}*
-
+  - python {{ PY_VER }}*
+  - epics-base {{ EPICS }}*
+  - numpy {{ NPY_VER }}*
```
Comment (tangkong) [inline conda-recipe/meta.yaml:24]: Note to self: figure out where this variable is set.

File: setup.py [hunk 2]
```diff
@@ -28,12 +27,4 @@
...
-
-
-setup(name='pyca',
-      version=versioneer.get_version(),
-      cmdclass=versioneer.get_cmdclass(),
-      description='python channel access library',
-      packages=['psp'],
-      ext_modules=[pyca],
-     )
+setup(ext_modules=[pyca,])
```
Comment (tangkong) [inline setup.py]: My cursory research suggests this is how to mix pyproject and C extensions, but I need to figure out a good way to test this

File: conda-recipe/meta.yaml (hunk 1 above)
Comment (ZLLentz) [reply conda-recipe/meta.yaml:24]: conda build has a bunch of built-in variables including this one: https://docs.conda.io/projects/conda-build/en/stable/user-guide/environment-variables.html#environment-variables-set-during-the-build-process

In this case, the recipe forces us to use a consistent NP version between the compilation step and the runtime

File: psp/__init__.py [hunk 3]
```diff
@@ -1,7 +1,8 @@
-__all__ = ['Pv','options']
+__all__ = ['Pv', 'options']
 
-from psp.Pv import Pv as PV
+from psp.Pv import Pv as PV  # noqa
```
Comment (ZLLentz) [inline psp/__init__.py]: This file has (always?) had a desync between the `__all__` and the import here

File: setup.py (hunk 2 above)
Comment (ZLLentz) [reply setup.py]: Were you able to compile pyca?

**Review by ZLLentz — COMMENTED** [review]
The changes all look correct
Did you manage to build and test it?
Is there any CI running here?

File: psp/__init__.py (hunk 3 above)
Comment (tangkong) [reply psp/__init__.py]: I had assumed some other dependency had `from psp import PV` and didn't want to break that.

Comment (ZLLentz) [reply psp/__init__.py]: But `__all__` should have `PV` instead of `Pv`, right?

Comment (ZLLentz) [reply psp/__init__.py]: Or does it as written give you the entire submodule on the `*` import anyway?

Comment (ZLLentz) [reply psp/__init__.py]: TIL
```
In [1]: from psp import *

In [2]: PV
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Input In [2], in <cell line: 1>()
----> 1 PV

NameError: name 'PV' is not defined

In [3]: Pv
Out[3]: <module 'psp.Pv' from '/cds/group/pcds/pyps/conda/py39/envs/pcds-5.8.4/lib/python3.9/site-packages/psp/Pv.py'>

In [4]: options
Out[4]: <module 'psp.options' from '/cds/group/pcds/pyps/conda/py39/envs/pcds-5.8.4/lib/python3.9/site-packages/psp/options.py'>

In [5]: caget
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Input In [5], in <cell line: 1>()
----> 1 caget

NameError: name 'caget' is not defined
```

Comment (ZLLentz) [reply psp/__init__.py]: Nothing to change here, I'm not well-versed in star imports apparently

File: setup.py (hunk 2 above)
Comment (tangkong) [reply setup.py]: I thought I did, but I tried again on a fresh environment and installation failed looking for numpy at setup.  (as this setup.py file needs numpy).

Even after installing numpy setuptools can't find it.  Weird

Comment (tangkong) [reply setup.py]: Ah ok I just need to do this properly through pyproject.toml.  (under build-system.requires)

Comment (tangkong) [reply setup.py]: This was more confusing than I thought, but with [cef4cd](https://github.com/slaclab/pyca/pull/18/commits/cef4cd6f22676323c4c90298bfa05e72bade47ab) we're able to install pyca, and run the test suite (given pytest is installed).  A quick spot check shows I can also import both `psp` and `pyca` in a python interpreter

**Comment by ZLLentz** [comment]
The changes all look good to me, I'm going to clone the branch and try to do a local conda build

**Comment by ZLLentz** [comment]
I needed to fight the recipe a bit because you need to have the _version.py generated prior to running conda build if the version file is references in the recipe file. We usually do this automatically in our standard ci over on pcdshub.
The build is going strangely slow through conda, I don't think it's even started doing anything yet...

**Comment by ZLLentz** [comment]
Actually, it seems to not work at all through conda...
```
  File "/cds/home/z/zlentz/miniconda3/lib/python3.9/site-packages/conda/models/match_spec.py", line 815, in merge
    raise ValueError(
ValueError: Incompatible component merge:
  - '*mpich*'
  - 'mpi_mpich_*'
```
I'll try a pip install into an existing env with python, epicscorelibs, and numpy maybe?

The conda recipe on conda forge has a pretty suspicious patch: https://github.com/conda-forge/pyca-feedstock/blob/main/recipe/0001-setup.patch

**Comment by ZLLentz** [comment]
pip installing works fine
runs with latest python+numpy

```
>>> pv2 = Pv("IM2K0:XTES:CAM:IMAGE1:ArrayData", use_numpy=True)
>>> pv2.get()
array([55, 58, 53, ...,  0,  0,  0], dtype=int16)
```

I'm going to somewhat suggest dumping all the conda references in this repo if we can't make them build trivially

**Comment by tangkong** [comment]
I admittedly didn't test the conda install yet, thanks for hitting your head against that for a bit.  On our end we'd need to switch this to a pip package if we dump the conda things.  I'll look at the conda recipe a bit tomorrow but I could be in favor of dropping conda support for now.

**Comment by ZLLentz** [comment]
We still do have a conda-forge so we wouldn't be dropping it entirely
I'm happy to look at it again with fresh eyes too

File: psp/__init__.py [hunk 4]
```diff
@@ -1,7 +1,3 @@
-__all__ = ['Pv','options']
+__all__ = ['Pv', 'options']
 
-from psp.Pv import Pv as PV
-
-from ._version import get_versions
-__version__ = get_versions()['version']
-del get_versions
```
Comment (ZLLentz) [inline psp/__init__.py:7]: one more small thing I noticed: since this versioneer stuff is gone, psp no longer self-reports it's version at the top-level

Comment (tangkong) [reply psp/__init__.py:7]: Ah right there's some `version.py` stuff to add that references the generated `_version.py`

**Comment by tangkong** [comment]
I went to see if applying the patch to the package would break the pip install, and it does.  (It's also using numpy calls that have been deprecated in py3.12)

We might consider referencing `epicscorelibs`'s example, which successfully builds DSOs for both pip and conda builds.  Their case is a bit strange though, they technically have a `pyproject.toml` but it only set up build requirements for the 600 line `setup.py` file.  
[epicscorelibs setup.py](https://github.com/mdavidsaver/epicscorelibs/blob/master/setup.py)
[epicscorelibs conda-forge](https://github.com/conda-forge/epicscorelibs-feedstock/blob/main/recipe/meta.yaml)

**Comment by tangkong** [comment]
Ok, after some discussion I think we can cut this effort short here.  This repo only ever supported pip installs anyway, with the conda-feedstock applying a whacky patch.  Moving forward for 3.12 we'll have to use pip, since that patch uses deprecated numpy methods.

**Review by ZLLentz — APPROVED** [review]
We might need another follow-up but we'll pick this back up in a couple weeks

**Pull request merged**